### PR TITLE
(partially) upgrade numpy

### PR DESCRIPTION
**Pull request by natestemen**
## Description

Upgrade to `numpy==1.21.6`. This requires adding new type hints for numpy arrays. 64-bit precision was chosen for all type specifications as 64-bit precision is python's default floating point precision. There are some places where I saw `np.complex128`. Is there a need for this high of precision?

I've done my best to try and understand where arrays should be int vs float vs complex, but it's hard to tell in some cases. I've also had to add `# type: ignore` comments in a few places as I have not been able to work out how to satisfy `mypy`. Most of these cases seem to be surrounding code where we are passing around random seeds that are optional.

---

fixes https://github.com/unitaryfund/mitiq/issues/1214

This upgrade will also close out the dependabot security alert related to [CVE-2021-33430](https://nvd.nist.gov/vuln/detail/CVE-2021-33430).

### CODE REVIEW

**Review by andreamari — APPROVED** [review]
Great job @natestemen ! 
LGTM up to a minor simplification. 

Note: There is an intrinsic inconsistency in Mitiq since a large part of the old code (e.g. factories) assume that expectation values are real. After introducing observables, we extended expectation values to complex numbers. I am surprised `mypy` is not complaining about this :smile: .  This is quite confusing and at some point it should be fixed in some way see e.g. #1254  .   Anyway this PR has a different scope, so it's good to merge it now.

**Comment by andreamari** [comment]
I tried a couple of changes in executor.evaluate() by I could not find any good solution to simplify `DensityMatrixLike`,  `MeasurementResultLike` etc..
So for me it's fine to revert the last commit such that we can merge this PR.

File: mitiq/executor/executor.py [hunk 1]
```diff
@@ -85,6 +87,9 @@ def __init__(
 
         executor_annotation = inspect.getfullargspec(executor).annotations
         self._executor_return_type = executor_annotation.get("return")
+        print(executor)
+        print(executor_annotation)
+        print(self._executor_return_type)
```
Comment (andreamari) [inline mitiq/executor/executor.py:92]:
```suggestion
```
This is probably the reason for the failing test

Comment (andreamari) [reply mitiq/executor/executor.py:92]: I'll self-commit this just to start CI tests.

**Review by nathanshammah — APPROVED** [review]
@natestemen looks great. I am not aware of the need of higher precision, so fine with complex64. With regards to the choice of float vs complex, it's mainly when a density matrix could be passed. We may spot later on that some typings need a switch, but since checks are passing and all looks good, we can merge. Thanks @andreamari for the bug fix.

**Pull request merged**